File: amm/compiler/interface/src/main/java/ammonite/compiler/iface/Parser.java
package ammonite.compiler.iface;

import java.util.Map;

public abstract class Parser {

  public abstract ParsedImportHooks importHooks(
    CodeSource source,
    Map.Entry<Integer, String>[] statements
  );

  public abstract ScriptBlock[] scriptBlocks(
    String rawCode,
    String fileName
  ) throws ScriptSplittingError; // Either[String, IndexedSeq[(String, Seq[String])]]

  public abstract ScriptBlock[] scriptBlocksWithStartIndices(
    String rawCode,
    String fileName
  ) throws ScriptSplittingError; // Either[Parsed.Failure, IndexedSeq[(Int, String, Seq[(Int, String)])]]

  public static final class ParsedImportHooks {
    private final String[] hookStatements;
    private final ImportTree[] importTrees;

    public ParsedImportHooks(
      String[] hookStatements,
      ImportTree[] importTrees
    ) {
      this.hookStatements = hookStatements;
      this.importTrees = importTrees;
    }

    public String[] hookStatements() {
      return hookStatements;
    }
    public ImportTree[] importTrees() {
      return importTrees;
    }

    // TODO equals, hashCode, toString
  }

  public static final class ScriptBlock {
    private final int startIndex;
    private final String ncomment;
    private final Map.Entry<Integer, String>[] codeWithStartIndices;

    public int startIndex() {
      return startIndex;
    }
    public String ncomment() {
      return ncomment;
    }
    public Map.Entry<Integer, String>[] codeWithStartIndices() {
      return codeWithStartIndices;
    }

    public ScriptBlock(
      String ncomment,
      Map.Entry<Integer, String>[] codeWithStartIndices
    ) {
      this.startIndex = 0;
      this.ncomment = ncomment;
      this.codeWithStartIndices = codeWithStartIndices;
    }

    public ScriptBlock(
      int startIndex,
      String ncomment,
      Map.Entry<Integer, String>[] codeWithStartIndices
    ) {
      this.startIndex = startIndex;
      this.ncomment = ncomment;
      this.codeWithStartIndices = codeWithStartIndices;
    }

    // TODO equals, hashCode, toString
  }

  public static class ScriptSplittingError extends Exception {
    private final int index;
    private final String expected;
    public ScriptSplittingError(String message) {
      super(message);
      this.index = -1;
      this.expected = null;
    }
    public ScriptSplittingError(
      String message,
      int index,
      String expected
    ) {
      super(message);
      this.index = index;
      this.expected = expected;
    }

    public int index() {
      return index;
    }
    public String expected() {
      return expected;
    }
  }
}

File: amm/compiler/interface/src/main/scala/ammonite/compiler/iface/Parser.scala
package ammonite.compiler.iface

import ammonite.util.ImportTree
import ammonite.util.Util.CodeSource

abstract class Parser {

  def split(
    code: String,
    ignoreIncomplete: Boolean = true,
    fileName: String = "(console)"
  ): Option[Either[String, Seq[String]]]

  final def parseImportHooks(
    source: CodeSource,
    stmts: Seq[String]
  ): (Seq[String], Seq[ImportTree]) =
    parseImportHooksWithIndices(source, stmts.map((0, _)))
  def parseImportHooksWithIndices(
    source: CodeSource,
    stmts: Seq[(Int, String)]
  ): (Seq[String], Seq[ImportTree])

  /**
    * Splits up a script file into its constituent blocks, each of which
    * is a tuple of (leading-whitespace, statements). Leading whitespace
    * is returned separately so we can later manipulate the statements e.g.
    * by adding `val res2 = ` without the whitespace getting in the way
    */
  def splitScript(
    rawCode: String,
    fileName: String
  ): Either[String, IndexedSeq[(String, Seq[String])]]

  def scriptBlocksWithStartIndices(
    rawCode: String,
    fileName: String
  ): Either[Parser.ScriptSplittingError, Seq[Parser.ScriptBlock]]

  def defaultHighlight(buffer: Vector[Char],
                       comment: fansi.Attrs,
                       `type`: fansi.Attrs,
                       literal: fansi.Attrs,
                       keyword: fansi.Attrs,
                       reset: fansi.Attrs): Vector[Char]

  def isObjDef(code: String): Boolean
}

object Parser {

  case class ParsedImportHooks(
    hookStatements: Seq[String],
    importTrees: Seq[ImportTree]
  )

  case class ScriptBlock(
    startIndex: Int,
    ncomment: String,
    codeWithStartIndices: Seq[(Int, String)]
  )

  object ScriptBlock {
    def apply(
      ncomment: String,
      codeWithStartIndices: Seq[(Int, String)]
    ): ScriptBlock =
      ScriptBlock(0, ncomment, codeWithStartIndices)
  }

  class ScriptSplittingError(
    message: String,
    val index: Int = -1,
    val expected: String = ""
  ) extends Exception(message)
}
